ledger/ledger-consensual/src/main/kotlin/net/corda/v5/ledger/consensual/ConsensualLedgerService.kt

```diff
@@ -21,4 +24,33 @@ interface ConsensualLedgerService {
     fun receiveFinality( ()-> ... )
     ... Vault ...
     */
+
+    /**
+     * Collects signatures, records and broadcasts to involved peers a [ConsensualSignedTransaction].
+     *
+     * @param signedTransaction The [ConsensualSignedTransaction] to finalise and recorded locally and with peer [sessions].
+     * @param sessions The [FlowSession]s of the peers involved in the transaction.
+     *
+     * @return The fully signed [ConsensualSignedTransaction] that was recorded.
+     */
+    @Suspendable
+    fun finality(
+        signedTransaction: ConsensualSignedTransaction,
+        sessions: List<FlowSession>
```

Comment:
We need to clarify a few things about this, because we have confusing sources of participants we need in finality. Ultimately we need 1. Those who need to sign (plus the implicit presence of the Notary) 2. Those who need to receive a copy reliably on completion.  3. Those who can see the states on vault insert. The transaction has a notary (PublicKey) and probably a service MemberX500Name. The transaction also has CommandAndSignatories with their required signers. FInally, ContractState has participants that controls visibility. However, note that the transaction is in terms of PublicKeys and FlowSessions is in terms of MemberX500Names (and not sure if you can get that from structure)

So we probably should have a chat about who and when to use these things

Reply:
I guess in C4 this works by each session participant evaluating what they feel OK with signing for i.e. identifying their keys from required signers and signing. Challenge there is that lookup of ones own keys by PublicKey/hash of PublicKey is not scalable when there are many accounts/CI keys. We might need to consider how during the negotiation phase peers make a note of what keys they intend to use with the transaction

Reply:
Just as an aside, I was idly wondering yesterday about a scheme where we work out what sessions we need internally in the finality flow, rather than ask the user to provide them. The advantage is that we should be able to verify up front if we're able to address the right network members, as we can either get the member X500 name from the key for well known keys, or come up with a secondary system flow for dealing with confidential keys further down the line. However, I think that might cause problems when thinking about inline vs non-inline responder flows, as that sort of idea would require non-inline. Perhaps some mixture of the two could work, where it's a pure verification step rather than actually making the sessions?

Reply:
Passing in the sessions is sticking with the Corda 4 approach - creating sessions under the hood is tricky for inline flows as mentioned.
I like the idea of checking that we can resolve the keys to names, and obviously we could then create the matching sessions which would give a nicer API.

In the Consensual ledger world, there is probably not much point to it - I would expect virtually all use cases to get X500 names for the required counterparties as input to the flow, and resolving those to keys for the signing, and sessions for the flow messaging.

I think from a UTOX user perspective, there are a couple of different use cases: 
- you might deal with known counterparties (think a simple cash on ledger model, where the only signatures required are mine and the receiver), and we'll have the X500 names as inputs tying together keys and sessions. In this case, resolving sessions from keys would be a waste of resources.
- But you might also deal with the case where the required signatories are defined by participants in an input state - there we need to resolve key -> X500 anyway, and it might be a lot cleaner to do it under the hood and not bother the user with this.